spec.md

```diff
@@ -3149,6 +3167,184 @@ The CO MUST implement the specified error recovery behavior when it encounters t
+| Invalid `snapshot_id` | 5 NOT_FOUND | Indicates that the snapshot specified was not found. | The caller should re-check that this object exists. |
+| Invalid `starting_offset` | 11 OUT_OF_RANGE | The starting offset exceeds the volume size. | The caller should specify a `starting_offset` less than the volume's size. |
+
+#### GetDelta RPC
+
+The plugin must implement this RPC if it provides the SnapshotMetadata service.
+
+```protobuf
+// The GetDeltaRequest message is used to solicit metadata on the data
+// ranges that have changed between two snapshots.
+message GetDeltaRequest {
+  // This is the identifier of the snapshot against which changes
+  // are to be computed.
+  string base_snapshot_id = 1;
+
+  // This is the identifier of a second snapshot in the same volume,
+  // created after the base snapshot.
+  string target_snapshot_id = 2;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed. It is intended
+  // to be used to continue a previously interrupted call.
+  // The plugins may round down this offset to the nearest alignment
+  // boundary based on the BlockMetadataType used.
+  uint64 starting_offset = 3;
+
+  // This is an optional parameter, and if non-zero it specifies the
+  // maximum number of tuples to be returned in each
+  // GetDeltaResponse message returned by the RPC stream.
+  // The plugin will determine an appropriate value if 0, and is
+  // always free to send less than the requested value.
+  uint32 max_results = 4;
+}
+
+// GetDeltaResponse messages are returned in gRPC stream.
+// Cumulatively, they provide information on the data ranges that
+// have changed between the base and target snapshots specified
+// in the GetDeltaRequest message.
+message GetDeltaResponse {
+  // This specifies the style used in the BlockMetadata sequence.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  BlockMetadataType block_metadata_type = 1;
+
+  // This returns the size of the underlying volume.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  uint64 volume_size_bytes = 2;
+
+  // This is list of data range tuples. The number of entries
+  // in this list is controlled by the max_results value in the
+  // GetDeltaRequest message.
+  repeated BlockMetadata block_metadata = 3;
+}
+
+```
+
+##### GetDelta Errors
+If the plugin is unable to complete the `GetDelta` call successfully it
+must return a non-OK gRPC code in the gRPC status.
+
+The following conditions are well defined:
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Missing or otherwise invalid argument | 3 INVALID_ARGUMENT | Indicates that a required argument field was not specified or an argument value is invalid | The caller should correct the error and resubmit the call. |
```

Comment:
Should there be an error indicating that the original volume was deleted and the snapshots can not be tracked back? As CSI Snapshots are independent objects, a SP may not be able to link the snapshot back to the original volume and compute the diff between two snapshots. This can be seen as a requirement on the CSI-driver implementations, but an error for it might be useful.

Reply:
Good point. There can also be other reasons why the metadata service can't compute the diff between the two snapshots. Can we introduce a more generic error here, instead of just one pertaining to invalid snapshot-to-volume mapping (due to missing volume)?

Reply:
I think `NOT_FOUND` would suffice.  Any error message can be embellished with additional details.